src/crypto/README.md

```diff
@@ -0,0 +1,275 @@
+
+When new crypto protocols are added, they will be added into their own
+`crypto_` `*.h` and `*.cc` files.
+
+## Helpful concepts
+
+Node.js currently uses OpenSSL to provide it's crypto substructure.
+(Some custom Node.js distributions -- such as Electron -- use BoringSSL
+instead.)
+
+This section aims to explain some of the utilities that have been
+provided to make working with the OpenSSL APIs a bit easier.
+
+### Pointer Types
+
+Most of the key OpenSSL types need to be explicitly freed when they are
+no longer needed. Failure to do so introduces memory leaks. To make this
+easier (and less error prone), the `crypto_util.h` defines a number of
+smart-pointer aliases that should be used:
+
+```cpp
+using X509Pointer = DeleteFnPtr<X509, X509_free>;
+using BIOPointer = DeleteFnPtr<BIO, BIO_free_all>;
+using SSLCtxPointer = DeleteFnPtr<SSL_CTX, SSL_CTX_free>;
+using SSLSessionPointer = DeleteFnPtr<SSL_SESSION, SSL_SESSION_free>;
+using SSLPointer = DeleteFnPtr<SSL, SSL_free>;
+using PKCS8Pointer = DeleteFnPtr<PKCS8_PRIV_KEY_INFO, PKCS8_PRIV_KEY_INFO_free>;
+using EVPKeyPointer = DeleteFnPtr<EVP_PKEY, EVP_PKEY_free>;
+using EVPKeyCtxPointer = DeleteFnPtr<EVP_PKEY_CTX, EVP_PKEY_CTX_free>;
+using EVPMDPointer = DeleteFnPtr<EVP_MD_CTX, EVP_MD_CTX_free>;
+using RSAPointer = DeleteFnPtr<RSA, RSA_free>;
+using ECPointer = DeleteFnPtr<EC_KEY, EC_KEY_free>;
+using BignumPointer = DeleteFnPtr<BIGNUM, BN_free>;
+using NetscapeSPKIPointer = DeleteFnPtr<NETSCAPE_SPKI, NETSCAPE_SPKI_free>;
+using ECGroupPointer = DeleteFnPtr<EC_GROUP, EC_GROUP_free>;
+using ECPointPointer = DeleteFnPtr<EC_POINT, EC_POINT_free>;
+using ECKeyPointer = DeleteFnPtr<EC_KEY, EC_KEY_free>;
+using DHPointer = DeleteFnPtr<DH, DH_free>;
+using ECDSASigPointer = DeleteFnPtr<ECDSA_SIG, ECDSA_SIG_free>;
+using HMACCtxPointer = DeleteFnPtr<HMAC_CTX, HMAC_CTX_free>;
+using CipherCtxPointer = DeleteFnPtr<EVP_CIPHER_CTX, EVP_CIPHER_CTX_free>;
+```
+
+Examples of these being used are pervasive through the `src/crypto` code.
+
+### `ByteSource`
+
+The `ByteSource` class is a helper utility representing a *read-only* byte
+array. Instances can either wrap external ("foreign") data sources, such as
+an `ArrayBuffer` (`v8::BackingStore`) or allocated data. If allocated data
+is used, then the allocation is freed automatically when the `ByteSource` is
+destroyed.
+
+### `ArrayBufferOrViewContents`
+
+The `ArrayBufferOfViewContents` class is a helper utility that abstracts
+`ArrayBuffer`, `TypedArray`, or `DataView` inputs and provides access to
+their underlying data pointers. It is used extensively through `src/crypto`
+to make it easier to deal with inputs that allow any `ArrayBuffer`-backed
+object.
+
+### `AllocatedBuffer`
+
+The `AllocatedBuffer` utility is defined in `allocated_buffer.h` and is not
+specific to `src/crypto`. It is used extensively within `src/crypto`, however,
+to hold allocated data that is intended to be output in response to various
+crypto functions.
```

Comment:
For now we need it. But this is one of the points I'd like to resolve. If we can transition sec/crypto to use the BackingStore now, that would be ideal to avoid the later change